albatross/models/gp.h

```diff
@@ -59,295 +65,252 @@ template <typename FeatureType> struct GaussianProcessFit {
+  // Here we efficiently only compute the diagonal of the posterior
+  // covariance matrix.
+  Eigen::MatrixXd explained = train_ldlt.solve(cross_cov);
+  Eigen::VectorXd explained_variance =
+      explained.cwiseProduct(cross_cov).array().colwise().sum();
+  Eigen::VectorXd marginal_variance = prior_variance - explained_variance;
+  return MarginalDistribution(pred, marginal_variance.asDiagonal());
 }
 
-template <typename FeatureType>
-inline JointDistribution predict_from_covariance_and_fit(
-    const Eigen::MatrixXd &cross_cov, const Eigen::MatrixXd &pred_cov,
-    const GaussianProcessFit<FeatureType> &model_fit) {
-  const Eigen::VectorXd pred = cross_cov.transpose() * model_fit.information;
-  auto ldlt = model_fit.train_ldlt;
-  Eigen::MatrixXd posterior_cov = ldlt.solve(cross_cov);
-  posterior_cov = cross_cov.transpose() * posterior_cov;
-  posterior_cov = pred_cov - posterior_cov;
-  return JointDistribution(pred, posterior_cov);
+inline JointDistribution
+gp_joint_prediction(const Eigen::MatrixXd &cross_cov,
+                    const Eigen::MatrixXd &prior_cov,
+                    const Eigen::VectorXd &information,
+                    const Eigen::SerializableLDLT &train_ldlt) {
+  const Eigen::VectorXd pred = gp_mean_prediction(cross_cov, information);
+  Eigen::MatrixXd explained_cov =
+      cross_cov.transpose() * train_ldlt.solve(cross_cov);
+  return JointDistribution(pred, prior_cov - explained_cov);
 }
 
-template <typename FeatureType, typename SubFeatureType = FeatureType>
-using SerializableGaussianProcess =
-    SerializableRegressionModel<FeatureType,
-                                GaussianProcessFit<SubFeatureType>>;
+/*
+ * This GaussianProcessBase will provide a model which is capable of
+ * producing fits and predicting for any FeatureType that is supported
+ * by the covariance function.  Sometimes, however, you may want to
+ * do some preprocessing of features since that operation could
+ * happen in order N time, instead of repeatedly preprocessing every
+ * time the covariance function is evaluated.  To do this you'll want
+ * to define a custom ImplType.  See test_models.cc for an example.
+ */
+template <typename CovFunc, typename ImplType>
+class GaussianProcessBase
+    : public ModelBase<GaussianProcessBase<CovFunc, ImplType>> {
 
-template <typename FeatureType, typename CovarianceFunction>
-class GaussianProcessRegression
-    : public SerializableGaussianProcess<FeatureType> {
-public:
-  typedef GaussianProcessFit<FeatureType> FitType;
-  typedef CovarianceFunction CovarianceType;
+  template <typename FitFeatureType>
+  using GPFitType = Fit<GaussianProcessBase<CovFunc, ImplType>, FitFeatureType>;
 
-  GaussianProcessRegression()
+public:
+  GaussianProcessBase()
       : covariance_function_(), model_name_(covariance_function_.get_name()){};
-  GaussianProcessRegression(CovarianceFunction &covariance_function)
+  GaussianProcessBase(CovFunc &covariance_function)
```

Comment:
Should this be `const CovFunc &covariance_function`? We're copying `covariance_function` anyways it looks like.

Reply:
yeah, good catch.